### ISSUE TITLE
4.6.0-2 not starting on AWS Linux 2

### ISSUE THREAD

**Issue by lukaszsamson** [zcash/zcash#5858]
### Describe the issue
Official linux tarbals are not starting on AWS Linux 2

### Can you reliably reproduce the issue?
Yes
#### If so, please list the steps to reproduce below:
1. Start a vm with officially supported AWS Linux 2 AMI
2. Try to run zcash

### Expected behaviour
Official binary tarball is working on supported OS. (AWS Linux 2 is based on centos 7 and is going to be supported til June 30, 2023 - see https://aws.amazon.com/amazon-linux-2/faqs/). AWS Linux 3(or 2022) is not yet released.

### Actual behaviour + errors
/opt/zcash/bin/zcashd: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by /opt/zcash/bin/zcashd)

### The version of Zcash you were using:
4.6.0-2

### Machine specs:
- OS name + version: AWS Linux 2

### Any extra information that might be useful in the debugging process.
Up until now the binary tarball for debian stretch worked fine on AWS Linux 2

**Comment by lukaszsamson** [comment]
Build from source is failing as well
```
Building boost...
clang++: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by clang++)
clang++: /lib64/libtinfo.so.5: no version information available (required by clang++)
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/tools/common.jam:1164: in common.match-command-output from module common
error: '"clang++" "-target" "x86_64-pc-linux-gnu" "-B/root/zcash/depends/x86_64-pc-linux-gnu/native/bin" "-stdlib=libc++" "-m64" --version' exited with error code 1
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/tools/clang-linux.jam:101: in get-full-version from module clang-linux
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/tools/clang-linux.jam:106: in get-short-version from module clang-linux
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/tools/clang-linux.jam:55: in clang-linux.init from module clang-linux
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/build/toolset.jam:44: in toolset.using from module toolset
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/tools/clang.jam:33: in clang.init from module clang
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/build/toolset.jam:44: in toolset.using from module toolset
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/build/project.jam:1104: in using from module project-rules
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/user-config.jam:1: in modules.load from module user-config
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/build-system.jam:255: in load-config from module build-system
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/build-system.jam:449: in load-configuration-files from module build-system
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/build-system.jam:607: in load from module build-system
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/kernel/modules.jam:294: in import from module modules
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/tools/build/src/kernel/bootstrap.jam:139: in boost-build from module
/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/boost-build.jam:17: in module scope from module

make: *** [/root/zcash/depends/work/build/x86_64-pc-linux-gnu/boost/1_78_0-73e0d991e71/./.stamp_built] Error 1
make: Leaving directory `/root/zcash/depends'
```

**Comment by str4d** [comment]
We have never supported AWS Linux 2 AFAIK. So this issue is a feature request to add a new supported OS, not a bug.

We do not support CentOS 7 (we do test against CentOS 8). We also just dropped support for Debian 9 in 4.6.0-2, so that's why you'll have encountered this issue.

My first suggestion would be to try upgrading libc, but I have no experience with AWS Linux 2, so I don't know how feasible that is.

**Comment by lukaszsamson** [comment]
@str4d I know AWS Linux 2 wasn't officially supported. My issue is that a patch release introduces breaking changes just before network update and deprecation of previous version. Can we get back Debian stretch build?

**Comment by str4d** [comment]
> My issue is that a patch release introduces breaking changes just before network update and deprecation of previous version.

We already planned to remove Debian 9 support in 4.7.0, which we had hoped to have out last week (but bugfixes during testing mean we'll likely instead have it out this week). The 4.6.0-2 hotfix release was just to tide users over so we had a release in the field, but even if it had maintained Debian 9 support (which we didn't feel was safe to do - see below), you'd just have encountered this a week later. It's unfortunate, but given that Debian 9 reaches end-of-support in just a few months (at the end of June), not unexpected.

> Can we get back Debian stretch build?

The reason we dropped Debian 9 support is because [LLVM did not provide Ubuntu 16.04 binaries for 13.0.1](https://github.com/llvm/llvm-project/issues/53892), so we instead compile with the Ubuntu 18.04 binaries which use a newer glibc version (corresponding to the error you are seeing both using the provided binaries and compiling from source). We couldn't stay on 13.0.0 (for which Ubuntu 16.04 binaries were provided) because 13.0.1 fixed [various bugs, including some serious miscompilations, that we had to assume could affect us](https://github.com/zcash/zcash/pull/5609#issuecomment-1055460378). I see that LLVM 14.0.0 similarly does not provide Ubuntu 16.04 binaries.

If you want to produce a build that works on Debian 9, you can revert to Clang 13.0.0 (by reverting https://github.com/zcash/zcash/commit/2f1fbcc81f64d2a75b380c7eda008aab7de97a92); this would be at your own risk.

**Comment by daira** [comment]
If you want to stay with Amazon Linux, I would suggest upgrading to the [Amazon Linux 2022 preview](https://docs.aws.amazon.com/linux/al2022/release-notes/all-packages.html). It has glibc 2.34 as documented [here](https://docs.aws.amazon.com/linux/al2022/release-notes/all-packages-al2022-20220308.html), which is more recent than Debian Bullseye so should stay sufficiently current for some time.

**Comment by gituser** [comment]
I agree with @lukaszsamson 
>My issue is that a patch release introduces breaking changes just before network update and deprecation of previous version. 

This should be a major version bump of zcash as you break it completely for end users.

Thanks for the tip with commit revert of 2f1fbcc81f64d2a75b380c7eda008aab7de97a92

seems to be building fine without that commit.

**Comment by daira** [comment]
The major version of zcashd is reserved for the network upgrade number. This update could have been called 4.7.0 but that version is already planned to be the version that supports NU5 on testnet.

In any case, using a different version number wouldn't have changed the decision to drop support for distributions using a version of glibc that is insufficient for the clang/LLVM 13 binary builds. As @str4d said, that was necessary for security and correctness reasons. (Technically, it would be possible to build clang/LLVM/libc++ from source but that would in practice make building much more difficult, error-prone, and time-consuming.)

**Comment by str4d** [comment]
> This should be a major version bump of zcash

`zcashd` does not use SemVer (and has never done so).

Our long-standing policy has been that incrementing the first integer in the version (a "major" version bump in SemVer) correspond to changes to the consensus rules (i.e. `zcashd 5.0.0` will follow the NU5 consensus rules on mainnet, whereas `zcashd 4.*` will not). That is, changes that will affect _all_ users in the Zcash ecosystem, not just (a subset of) `zcashd` users.

That implicitly requires breaking changes to the `zcashd` software itself to be deployed by incrementing the second-or-later version integer. We've been somewhat inconsistent in the far past (we originally used the second version integer to indicate new non-circuit consensus rules), but since v3.0.0 in May 2020, we've consistently used second-integer increments ("minor" bumps in SemVer) for regular releases, in which we will make breaking changes that affect (subsets of) `zcashd` users, but not the entire Zcash ecosystem. The removal of Debian 9 support was intended to be deployed in v4.7.0, consistent with this.

Meanwhile, we have always reserved the right to deploy any kind of security-relevant change in hotfix releases; 4.6.0-2 was a hotfix release, not a patch release (which we rarely create). We decided that the Clang 13.0.1 dependency increase was security-relevant.

> as you break it completely for end users.

We had already planned to drop Debian 9 once it reached EOS at the end of June 2022, so the breakage this issue relates to was going to happen anyway (because AWS Linux 2 will be supported by Amazon until the end of June 2023, and it was not a platform we support).

We merge a lot of changes into this codebase, and while we make a best effort to convey the magnitude to which those changes affect users, we can't be expected to treat every possible change as breaking any `zcashd` user, especially on unsupported platforms. As you have confirmed, `zcashd` does indeed still run on your platform; all that changed was the build process we support stopped working.

The only end users for whom the 4.6.0-2 release meant a breaking change in support are Debian 9 users, and given that Debian 9 was released in 2017 and the 2022 EOS is only a few months away, we didn't miss by very much. I'd have loved to keep supporting Debian 9 right through to its EOS, but as mentioned above, we rely on the binaries that LLVM publishes.

**Comment by lukaszsamson** [comment]
> I would suggest upgrading to the [Amazon Linux 2022 preview](https://docs.aws.amazon.com/linux/al2022/release-notes/all-packages.html). I

There is no upgrade path yet. AWS promises it will be available when AWS Linux 2022 is out of preview.

Just for the record I ended up using docker image as it is least hacky and more future proof.

**Comment by gituser** [comment]
@lukaszsamson you can also build zcash without commit https://github.com/zcash/zcash/commit/2f1fbcc81f64d2a75b380c7eda008aab7de97a92

**Comment by daira** [comment]
Please note, for anyone else reading, that building without 2f1fbcc81f64d2a75b380c7eda008aab7de97a92 is not a supported or recommended configuration because [clang 13.0.1 fixes compilation bugs](https://github.com/zcash/zcash/pull/5609#issuecomment-1055287796), and we cannot be confident that these bugs don't affect zcashd. They include, for example, bugs that have caused [miscompilation of cryptographic code](https://github.com/llvm/llvm-project/issues/51299) in other projects.

**Comment by gituser** [comment]
@daira, yet you build zcash for Windows using old `clang v13.0.1`.

**Comment by daira** [comment]
The bugs in question, including the cryptographic code miscompilation, were [fixed in clang 13.0.1](https://github.com/llvm/llvm-project/issues/51299#issuecomment-981046826).

**Comment by daira** [comment]
Bugs in the clang 14.0.x milestones ([14.0.0](https://github.com/llvm/llvm-project/milestone/3?closed=1), [14.0.1 closed](https://github.com/llvm/llvm-project/milestone/5?closed=1), [14.0.1 open](https://github.com/llvm/llvm-project/milestone/5)) are mostly crash bugs (i.e. if the build succeeds then we haven't hit the bug) or specific to architectures we don't support; not miscompilations on supported architectures.

I can see an exception that only affects floating-point code ([original report](https://developercommunity.visualstudio.com/t/Prev-Issue---with-__assume-isnan-/1597317), [LLVM issue](https://github.com/llvm/llvm-project/issues/54547)), and likely does not affect us.

Other possible exceptions are:
* https://github.com/llvm/llvm-project/issues/51617 for which I think a fix was backported in https://github.com/llvm/llvm-project/issues/53885
* https://reviews.llvm.org/D115955 for which a fix was backported in https://github.com/llvm/llvm-project/issues/53678
* https://github.com/llvm/llvm-project/issues/53726 fixed in https://github.com/llvm/llvm-project/commit/d1387a26a54228da8cc275bba5f428b16676bc61

**Comment by str4d** [comment]
Also, the only reason we hadn't upgraded to Clang 14 is because we want to keep Clang and Rust's LLVM backend versions in sync. Rust 1.60 was released during the 4.7.0 RC period using LLVM 14, so we'll be migrating to that and Clang 14 right after the 4.7.0 release.

**Comment by gituser** [comment]
Thanks @daira and @str4d your explanations are very appreciated.

I've ended up updating my zcash containers to Ubuntu 18.04, so far so good `v4.7.0` is running fine.
